content/posts/maths/NMPC-Two-Wheel-Bicycle-Model.md

```diff
@@ -0,0 +1,186 @@
+\ &\mathbf{u}^i(k) \in \bm{\mu},\ k \in \Z_{[0, N-1]}, \\
+\ &v_x^i(k) \in \bm{\gamma},\ k \in \Z_{[0, N-1]}, \\
+\ &\mathbf{Q}_1 \in \mathbb{S}^{2}_{++}, \\
+\ &\mathbf{Q}_2 \in \mathbb{S}^{n_u}_{++}, \text{ and}\\
+\ &i \in \Z_{[0, \bm{S}]}.
+\end{aligned}
+$$
+</p>
+
+**Note:** The $i$ in the superscript of $\mathbb{P}^i_N$ is indicative of a formulation that can change over iterations, which will be addressed in the coming sections.
+
+Where, $i$ is the number of the iteration in $\bm{S}$ simulation steps, with $\mathbf{u}^0(-1)$ being the input that the system began with i.e., the input at $t=-T$. The idea of NMPC is to begin with iteration number $i=0$ at $t=0$ and solve $\mathbb{P}^0_N(\mathbf{z}^0(0))$ for ${}^\star \mathbf{U}^{0}$, where ${}^\star \mathbf{U}^i = \underset{\mathbf{u}^i}{\text{argmin}}\ \mathbb{P}^i_N(\mathbf{z}^i(0))$. Then, take the first computed input ${}^\star\mathbf{u}^0(0)$, apply it to the system and take a measurement of the resulting state ${}^\star\mathbf{z}^0(1)$. With $\mathbf{u}^1(-1) = {}^\star\mathbf{u}^0(0)$ and $\mathbf{z}^1(0) = {}^\star\mathbf{z}^0(1)$, proceed to iteration $i=1$ at $t=T$ to solve $\mathbb{P}^1_N(\mathbf{z}^1(0))$, obtain ${}^\star\mathbf{u}^1(0)$, apply it to the system to get ${}^\star\mathbf{z}^1(1)$ and move on to iteration $i=2$ at $t = 2T$. This procedure is repeated till $i=\bm{S}$.
+
+**Note:** The reader might notice that this iterative method only works assuming that the computation of ${}^\star \mathbf{U}^i$ takes less than $T$ seconds during each iteration. It is indeed the case and it is not clear what happens if each computation takes longer than $T$ seconds. This hurts the confidence on the designed system, as it might produce unexpected results. Thus, the selection of $T$ shall be of careful consideration owing to the trade off between the accuracy of the discrete approximation and computational limitations. 
+
+This setup is implemented and simulated in Python3 with $N=50,\ \bm{S} = 300,\ T=0.01 \text{ s},\ \mathbf{z}^0(0) = 0_{n_x \times 1} \text{ and }\mathbf{p}^i_d = [5, 5]^\mathsf{T}\ \forall\ i \in \Z_{[0, \bm{s}]}$. The following GIF shows the simulation results.
+
+<!-- <div>
+<img width="640" height="480" controls>
+  <source src="/mpc_car_st_slope_no_brake.gif" type="image/gif">
+</img>
+</div> -->
+<img src="/mpc_car_st_slope_no_brake.gif" alt="Simulation results" width="640" height="400">
+
+
+As it can be seen in the GIF, the NMPC controlled vehicle indeed drove towards (close to) the destination $[5, 5]^\mathsf{T}$, but it didn't converge. It could have been a matter of tuning $\mathbf{Q}_1$ and $\mathbf{Q}_2$, but before tuning them, the vehicle dynamics need to be modified to add a braking system. Remember, the end goal is to have an NMPC controlled autonomous vehicle that is *road safe*. These preliminary results are good enough for now as the weight matrices $\mathbf{Q}_1$ and $\mathbf{Q}_2$ might need retuning when the dynamics have changed.
```

Comment:
Did it not converge because the NMPC does not have any terminal ingredients or because you didn't let us run long enough?

Reply:
That may also be because it doesn't have proper terminal constraints. But at that time, I thought this could have been a result of improper tuning because it was doing **bang-bang** control.  I did experiment with the tuning at a later point, which mitigated the rapid switching between inputs and drove the vehicle much closer to the destination (including the '0' heading). But by that point, I had already added the obstacle avoidance features and since the goal for section 1 is to see if the formulation works, I didn't bother tuning. 

I did let it run longer during the initial formulation, but all it did was accelerate again and spiral towards the reference point. This continued for a few more iterations until it crashed (I didn't look for the error code here, probably should have).

Now that you have mentioned it, it might have been more likely due to the improper terminal constraints.